pinot-query-planner/src/main/java/org/apache/pinot/calcite/rel/rules/PinotAggregateToSemiJoinRule.java

```diff
@@ -50,18 +48,9 @@ public class PinotAggregateToSemiJoinRule extends RelOptRule {
       new PinotAggregateToSemiJoinRule(PinotRuleUtils.PINOT_REL_FACTORY);
 
   public PinotAggregateToSemiJoinRule(RelBuilderFactory factory) {
-    super(operand(LogicalAggregate.class, any()), factory, null);
-  }
-
-  @Override
-  @SuppressWarnings("rawtypes")
-  public boolean matches(RelOptRuleCall call) {
-    final Aggregate topAgg = call.rel(0);
-    if (!PinotRuleUtils.isJoin(topAgg.getInput())) {
-      return false;
-    }
-    final Join join = (Join) PinotRuleUtils.unboxRel(topAgg.getInput());
-    return PinotRuleUtils.isAggregate(join.getInput(1));
+    super(operand(Aggregate.class,
+            some(operand(Join.class, some(operand(RelNode.class, any()), operand(Aggregate.class, any()))))), factory,
+        null);
```

Comment:
Nit: I would recommend to start using the new way to instantiate rules instead of this one, which is deprecated. See PinotImplicitTableHintRule in https://github.com/apache/pinot/pull/13943.

Reply:
Yes, I also noticed that. `PinotSortExchangeCopyRule` is also using the config in the constructor (I believe the benefit is to allow reusing the same rule to match different trees).
We can use a separate PR to clean this up.

Reply:
I've recently ended up reading Hive's [CalcitePlanner class](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java), which is pretty similar to our QueryEnvironment class and I think that is something we can get inspiration of once our rule logic starts to be more complex.